### PR TITLE
sql: detect repeat overflow

### LINKED ISSUE #8042: sql: panic: runtime error: makeslice: len out of range
SQL:

```
SELECT repeat('string', 1797937339169437232);
```

Found with RSG.

### PR DESCRIPTION
**Pull request by madelynnblue**
Fixes #8042

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/8073)

### CODE REVIEW

File: sql/parser/builtins.go [hunk 1]
```diff
@@ -216,12 +216,20 @@ var Builtins = map[string][]Builtin{
 		Builtin{
 			Types:      ArgTypes{TypeString, TypeInt},
 			ReturnType: TypeString,
-			fn: func(_ *EvalContext, args DTuple) (Datum, error) {
+			fn: func(_ *EvalContext, args DTuple) (d Datum, err error) {
```
Comment (tamird) [inline sql/parser/builtins.go]: s/d/_/

File: sql/parser/builtins.go [hunk 2]
```diff
@@ -216,12 +216,20 @@ var Builtins = map[string][]Builtin{
...
 				s := string(*args[0].(*DString))
 				count := int(*args[1].(*DInt))
 				if count < 0 {
 					count = 0
 				}
+				// Repeat can overflow if len(s) * count is very large. The computation
```
Comment (tamird) [inline sql/parser/builtins.go]: FWIW, here's what PG does:
```
ERROR:  function repeat(unknown, bigint) does not exist
LINE 1: SELECT repeat('s', 9223372036854775807);
               ^
HINT:  No function matches the given name and argument types. You might need to add explicit type casts.
```

**Comment by mjibson** [comment]
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[sql/parser/builtins.go, line 219 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8073#-KNiEwg3SrP6mbktO44H-r1-219:-KNju9OLHRqQvN75CdtS:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4f8f088f4cc11629fa35260511770b76e8a2960/sql/parser/builtins.go#L219)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> s/d/_/
</details>
Done.

---

*[sql/parser/builtins.go, line 225 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8073#-KNiEwg3SrP6mbktO44H-r1-225:-KNjuB5k4ewmsbtzh4tW:1400627867) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4f8f088f4cc11629fa35260511770b76e8a2960/sql/parser/builtins.go#L225)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> FWIW, here's what PG does:
> 
> ```
> 
> ERROR:  function repeat(unknown, bigint) does not exist
> 
> LINE 1: SELECT repeat('s', 9223372036854775807);
> 
>                ^
> 
> HINT:  No function matches the given name and argument types. You might need to add explicit type casts.
> 
> ```
</details>
They must use 32-bit ints. Here's what happens when count is not a bigint:

```
postgres=# select repeat('asoentuhaosenuthaosnetuh', 2000000000);
ERROR:  requested length too large
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8073)*

**Comment by tamird** [comment]
Reviewed 1 of 1 files at r2.
Review status: 1 of 2 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[sql/parser/builtins.go, line 225 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8073#-KNiEwg3SrP6mbktO44H-r1-225:-KNkok6vXzzEVMP2Hek7:2041293688) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4f8f088f4cc11629fa35260511770b76e8a2960/sql/parser/builtins.go#L225)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary>
> They must use 32-bit ints. Here's what happens when count is not a bigint:
> 
> ```
> postgres=# select repeat('asoentuhaosenuthaosnetuh', 2000000000);
> ERROR:  requested length too large
> ```
</details>
I still think I'd prefer we used an artificial limit - leaving this to "whatever `make` can allocate" is worrying because an unprivileged user can OOM the server. MaxInt32 should certainly suffice?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8073)*

**Comment by RaduBerinde** [comment]
Review status: 1 of 2 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[sql/parser/builtins.go, line 225 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8073#-KNiEwg3SrP6mbktO44H-r1-225:-KNqbFU1QNwJ2cboVpz3:-1275744080) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4f8f088f4cc11629fa35260511770b76e8a2960/sql/parser/builtins.go#L225)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> I still think I'd prefer we used an artificial limit - leaving this to "whatever `make` can allocate" is worrying because an unprivileged user can OOM the server. MaxInt32 should certainly suffice?
</details>
What limit does PG impose? If it's a constant value we can use the same.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8073)*

**Comment by tamird** [comment]
Reviewed 1 of 2 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[sql/parser/builtins.go, line 225 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8073#-KNiEwg3SrP6mbktO44H-r1-225:-KNquVRqlBx9xuHTTlHX:1088360494) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4f8f088f4cc11629fa35260511770b76e8a2960/sql/parser/builtins.go#L225)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> What limit does PG impose? If it's a constant value we can use the same.
</details>
I think we're gonna need to read the source to find out.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8073)*

**Comment by RaduBerinde** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[sql/parser/builtins.go, line 225 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8073#-KNiEwg3SrP6mbktO44H-r1-225:-KNr1aV1EBYoOPr8bivy:-1375386643) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4f8f088f4cc11629fa35260511770b76e8a2960/sql/parser/builtins.go#L225)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> I think we're gonna need to read the source to find out.
</details>
I checked it out, a MaxAllocSize hard limit of (1GB - 1) is used. Then the memory allocation (`palloc`) uses a context-specific limit, and if that is exceeded the current context is killed and we error out with "out of memory".

The way they kill the current context is using something like a`throw`, except it's in C. Wow. http://doxygen.postgresql.org/elog_8c.html#a5fec6346bc9a7260191ea5ad8524b46f

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8073)*

**Comment by mjibson** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[sql/parser/builtins.go, line 225 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8073#-KNiEwg3SrP6mbktO44H-r1-225:-KO6NtBDSmNalCWJs9C2:-699785256) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4f8f088f4cc11629fa35260511770b76e8a2960/sql/parser/builtins.go#L225)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> I checked it out, a MaxAllocSize hard limit of (1GB - 1) is used. Then the memory allocation (`palloc`) uses a context-specific limit, and if that is exceeded the current context is killed and we error out with "out of memory".
> 
> The way they kill the current context is using something like a`throw`, except it's in C. Wow. http://doxygen.postgresql.org/elog_8c.html#a5fec6346bc9a7260191ea5ad8524b46f
</details>
Until we figure out a way to limit memory per-connection, I'm not sure this is much more we can do. I think this PR should be merged because it's better than what we have now (it prevents a panic), doesn't add any negative effects, and doesn't have an obvious better solution.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8073)*

**Comment by mjibson** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[sql/parser/builtins.go, line 225 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8073#-KNiEwg3SrP6mbktO44H-r1-225:-KO6OF0n7i9aNIUMdncf:-619751158) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4f8f088f4cc11629fa35260511770b76e8a2960/sql/parser/builtins.go#L225)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary>
> Until we figure out a way to limit memory per-connection, I'm not sure this is much more we can do. I think this PR should be merged because it's better than what we have now (it prevents a panic), doesn't add any negative effects, and doesn't have an obvious better solution.
</details>
s/this is/there is

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8073)*

**Comment by tamird** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[sql/parser/builtins.go, line 225 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8073#-KNiEwg3SrP6mbktO44H-r1-225:-KO6Pwr0IT1CvMTJp7L7:-284773538) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4f8f088f4cc11629fa35260511770b76e8a2960/sql/parser/builtins.go#L225)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary>
> s/this is/there is
</details>
I would limit it to 1MB for now. That seems obviously better?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8073)*

**Comment by RaduBerinde** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[sql/parser/builtins.go, line 225 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8073#-KNiEwg3SrP6mbktO44H-r1-225:-KO6RPZfF6-DZaZjpIoN:-201109668) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4f8f088f4cc11629fa35260511770b76e8a2960/sql/parser/builtins.go#L225)):*
> Until we figure out a way to limit memory per-connection, I'm not sure this is much more we can do.
Ah, absolutely. Just wanted to write down what I found about PG. Whatever infrastructure we will add for memory management will need to take care of this (and will probably have a hard allocation limit as well).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8073)*

**Comment by mjibson** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[sql/parser/builtins.go, line 225 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8073#-KNiEwg3SrP6mbktO44H-r1-225:-KO6RTykfnbD6dYUi4R_:1605487541) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4f8f088f4cc11629fa35260511770b76e8a2960/sql/parser/builtins.go#L225)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> > Until we figure out a way to limit memory per-connection, I'm not sure this is much more we can do.
> Ah, absolutely. Just wanted to write down what I found about PG. Whatever infrastructure we will add for memory management will need to take care of this (and will probably have a hard allocation limit as well).
</details>
The point of this PR is to prevent this specific panic. If we want to do any memory-related work, that should happen in another PR. Even with the 1MB limit, users can still do:

```
SELECT repeat('s', 1e9), repeat('s', 1e9), repeat('s', 1e9), ...
```

But I'm not trying to solve that problem yet.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8073)*

**Comment by tamird** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[sql/parser/builtins.go, line 225 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8073#-KNiEwg3SrP6mbktO44H-r1-225:-KO6XdaPfhkNKuWlhr27:944015337) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4f8f088f4cc11629fa35260511770b76e8a2960/sql/parser/builtins.go#L225)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary>
> The point of this PR is to prevent this specific panic. If we want to do any memory-related work, that should happen in another PR. Even with the 1MB limit, users can still do:
> 
> ```
> SELECT repeat('s', 1e9), repeat('s', 1e9), repeat('s', 1e9), ...
> ```
> 
> But I'm not trying to solve that problem yet.
</details>
Uh, okay. If you want to add a limit in another PR, that's fine, but it seems convenient to do it here. Can we agree that a limit would be a good thing, even if it was set aggressively low, perhaps lower than 1MB?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8073)*

**Comment by RaduBerinde** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[sql/parser/builtins.go, line 225 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8073#-KNiEwg3SrP6mbktO44H-r1-225:-KO6YnuTc-MXrGBXuBFP:1406128313) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4f8f088f4cc11629fa35260511770b76e8a2960/sql/parser/builtins.go#L225)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Uh, okay. If you want to add a limit in another PR, that's fine, but it seems convenient to do it here. Can we agree that a limit would be a good thing, even if it was set aggressively low, perhaps lower than 1MB?
</details>
Once we have memory management, having an artificial limit will no longer make sense. Unless we have a good reason, I say we should avoid peppering the code with arbitrary limits in random places that will need to be found and removed later.

If we need to add one, it should be associated with an issue (so we can easily find it later).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8073)*

**Comment by tamird** [comment]
OK.

**Pull request merged**